### ISSUE TITLE
SimpleDeserializers.java does not implement findDesererializer as expected.

### ISSUE THREAD

**Issue by parthshyara** [FasterXML/jackson-databind#2684]
TLDR; SimpleDeserializers.java is not analogous to SimpleSerializers.java

Suppose, we have registered a custom serializer and deserialiizer against `Map.class`. We have a custom Map that implements Map interface. While serialization of this custom Map, the custom serializer is picked because `_interfaceMappings` are configured in SimpleSerializers.java. Similar concept seems to be missing in SimpleDeserializers.java and hence, while deserializing the custom Map, the registered custom deserializer is not getting picked. This is unexpected.

Probable fix: Implement `findDeserializer` that uses `_interfaceMappings` to resolve deserializer for super-interfaces and super-class at runtime instead of `_find` that uses only `_classMappings` in SimpleDeserializers.java.

**Comment by cowtowncoder** [comment]
No, I don't think so. Map and Collection deserializers can not work properly without getting type parameterization, and you really should not register them using `SimpleDeserializers`.

This is different from serialization side where one may choose to only use runtime (dynamic) types of values.

**Comment by parthshyara** [comment]
Your made a valid point. But in a certain use case, I have written custom (de)serializer for Map type and registered it in the ObjectMapper. In the json, I have properly stored type information while serializing so that the deserialization works correctly. I am explicitly using `readValue(..., Map.class)` if the deserializing class is assignable from Map as a workaround. But in a custom Java Object, if a Map variable is present, and I use this objectmapper, it serializes the maps to custom json due to dynamic resolution of class/interface and fails while deserialization because of non-analogous behavious of SimpleDeserializer. So, shouldn’t the behaviour be expected to be identical?
Moreover, `_classMappings` and `_interfaceMappings` will only be populated by add(De)Serializer() methods.

**Comment by cowtowncoder** [comment]
Regardless, the choice of not handling "non-simple" type with `SimpleDeserializers` is intentional and I will not change that. If you need to register such types, implement custom `Deserializers` instead. I am open to changes to Javadoc stating the limitation, but not to changing behavior.
